### PR TITLE
[SPARK-9819][Streaming][Documentation] Clarify doc for invReduceFunc in incremental versions of reduceByWindow

### PR DESCRIPTION
**Pull request by huitseeker**
- that reduceFunc and invReduceFunc should be associative
- that the intermediate result in iterated applications of inverseReduceFunc
  is its first argument

### CODE REVIEW

File: python/pyspark/streaming/dstream.py [hunk 1]
```diff
@@ -454,7 +454,10 @@ def reduceByWindow(self, reduceFunc, invReduceFunc, windowDuration, slideDuratio
         This is more efficient than `invReduceFunc` is None.
 
         @param reduceFunc:     associative reduce function
-        @param invReduceFunc:  inverse reduce function of `reduceFunc`
+        @param invReduceFunc:  inverse reduce function of `reduceFunc`; such that for all x,
+                               and invertible y:
+                               `invReduceFunc(reduceFunc(x, y), y) = x`
+                               Both `reduceFunc` and `invReduceFunc` should be associative.
```
Comment (tdas) [inline python/pyspark/streaming/dstream.py]: Are you sure about this? The inverse reduce of X+Y is Z-X which is not associative.
(a - b) - c =/= a - (b - c)

Comment (huitseeker) [inline python/pyspark/streaming/dstream.py]: Correct, only `reduceFunc` should be

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala [hunk 2]
```diff
@@ -336,7 +336,8 @@ class JavaPairDStream[K, V](val dstream: DStream[(K, V)])(
    * However, it is applicable to only "invertible reduce functions".
    * Hash partitioning is used to generate the RDDs with Spark's default number of partitions.
    * @param reduceFunc associative reduce function
-   * @param invReduceFunc inverse function
+   * @param invReduceFunc inverse function; such that for all x, invertible y:
+   *                      `invReduceFunc(reduceFunc(x, y), y) = x`
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]: Can you also add `invReduceFunc(reduceFunc(x, y), x) = y`

Comment (huitseeker) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]: I'm afraid this one would be false as it implies commutativity. Try with string concatenation.

**Comment by JoshRosen** [comment]
@tdas, what's the status on this PR? I'm not familiar with the documentation being changed here, but this looks like a small PR that might be very close to being merged, so I just wanted to bump it.

**Comment by zsxwing** [comment]
LGTM. ping @tdas to take a final look

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala (hunk 2 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]: Why not
reduceFunc("x", "y") = "xy"   ... y is always added to right
inverseReduceFunc("xy", "x") = "y"    ... x is always removed from left

**Comment by tdas** [comment]
Sorry i forgot about this PR completely. Just one more nit that i commented on.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala (hunk 2 above)
Comment (huitseeker) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]:
```
Welcome to Scala version 2.11.7 (OpenJDK 64-Bit Server VM, Java 1.7.0_91).
Type in expressions to have them evaluated.
Type :help for more information.

scala> def f(s1: String, s2: String) = s1 + s2
f: (s1: String, s2: String)String

scala> def fInv(s1: String, s2: String) = s1.stripSuffix(s2)
fInv: (s1: String, s2: String)String

scala> val x = "aaa"
x: String = aaa

scala> val y = "bbb"
y: String = bbb

scala> fInv(f(x, y), y)
res0: String = aaa

scala> fInv(f(x, y), y) === x
<console>:15: error: value === is not a member of String
       fInv(f(x, y), y) === x
                        ^

scala> fInv(f(x, y), y) == x
res2: Boolean = true

scala> fInv(f(x, y), x)
res3: String = aaabbb

scala> fInv(f(x, y), x) === y
<console>:15: error: value === is not a member of String
       fInv(f(x, y), x) === y
                        ^

scala> fInv(f(x, y), x) == y
res5: Boolean = false
```

**Comment by huitseeker** [comment]
@tdas Answered.

**Comment by holdenk** [comment]
Hi @huitseeker I was going through some old PRs and I noticed this doc change, its got some conflicts with master but could we maybe update it and see if @tdas still thinks it would be a good improvement? Always nice to improve docs :)

**Comment by huitseeker** [comment]
@holdenk @tdas Done, thanks for the heads up  !

**Comment by zsxwing** [comment]
ok to test

**Comment by zsxwing** [comment]
LGTM. Merging to master, 2.0 and 1.6.

**Comment by zsxwing** [comment]
okey. Didn't push to 1.6 due to the conflicts. Thanks, @huitseeker

**Pull request closed without merging**